cache/refs.go

```diff
@@ -308,25 +334,260 @@ func (sr *immutableRef) Info() RefInfo {
 	}
 }
 
-func (sr *immutableRef) Extract(ctx context.Context) error {
-	_, err := sr.sizeG.Do(ctx, sr.ID()+"-extract", func(ctx context.Context) (interface{}, error) {
+func (sr *immutableRef) ociDesc() (ocispec.Descriptor, error) {
+	desc := ocispec.Descriptor{
+		Digest:      digest.Digest(getBlob(sr.md)),
+		Size:        getBlobSize(sr.md),
+		MediaType:   getMediaType(sr.md),
+		Annotations: make(map[string]string),
+	}
+
+	diffID := getDiffID(sr.md)
+	if diffID != "" {
+		desc.Annotations["containerd.io/uncompressed"] = diffID
+	}
+
+	createdAt := GetCreatedAt(sr.md)
+	if !createdAt.IsZero() {
+		if createdAt, err := createdAt.MarshalText(); err != nil {
+			return ocispec.Descriptor{}, err
+		} else {
+			desc.Annotations["buildkit/createdat"] = string(createdAt)
+		}
+	}
+
+	return desc, nil
+}
+
+// order is from parent->child, sr will be at end of slice
+func (sr *immutableRef) parentRefChain() []*immutableRef {
+	var count int
+	for ref := sr; ref != nil; ref = ref.parent {
+		count++
+	}
+	refs := make([]*immutableRef, count)
+	for i, ref := count-1, sr; ref != nil; i, ref = i-1, ref.parent {
+		refs[i] = ref
+	}
+	return refs
+}
+
+func (sr *immutableRef) GetRemote(ctx context.Context, createIfNeeded bool, compressionType compression.Type) (*Remote, error) {
+	ctx, done, err := leaseutil.WithLease(ctx, sr.cm.LeaseManager, leaseutil.MakeTemporary)
+	if err != nil {
+		return nil, err
+	}
+	defer done(ctx)
+
+	err = sr.computeBlobChain(ctx, createIfNeeded, compressionType)
+	if err != nil {
+		return nil, err
+	}
+
+	mprovider := contentutil.NewMultiProvider(nil)
+	remote := &Remote{
+		Provider: mprovider,
+	}
+
+	for _, ref := range sr.parentRefChain() {
+		desc, err := ref.ociDesc()
+		if err != nil {
+			return nil, err
+		}
+
+		// NOTE: The media type might be missing for some migrated ones
+		// from before lease based storage. If so, we should detect
+		// the media type from blob data.
+		//
+		// Discussion: https://github.com/moby/buildkit/pull/1277#discussion_r352795429
+		if desc.MediaType == "" {
+			desc.MediaType, err = compression.DetectLayerMediaType(ctx, sr.cm.ContentStore, desc.Digest, false)
+			if err != nil {
+				return nil, err
+			}
+		}
+
+		remote.Descriptors = append(remote.Descriptors, desc)
+
+		provider, err := ref.getProvider(ctx, sr.descHandler)
+		if err != nil {
+			return nil, err
+		}
+		mprovider.Add(desc.Digest, provider)
+	}
+	return remote, nil
+}
+
+func (sr *immutableRef) getProvider(ctx context.Context, dh *DescHandler) (content.Provider, error) {
+	if isLazy, err := sr.isLazy(ctx); !isLazy {
```

Comment:
I previously looked through the underlying implementation of the content store methods it's calling and concluded the synchronization wasn't necessary here, but I'll go through it all again to double-check and get back

Reply:
Okay yeah, I still believe this basically works as is. There's a few relevant pieces of code:
1. We always use [`contentutil.Copy` when de-lazying, which in turn uses a `FetchHandler`](https://github.com/sipsma/buildkit/blob/94e8ab5a3b104d6ba7884375e2d8f391ceb3b1e5/util/contentutil/copy.go#L16)
2. The `FetchHandler` [internally opens a content writer with a writer ref consistent for a given oci descriptor](https://github.com/containerd/containerd/blob/334f567e0130b2734a3a5439f2bb644d14aa66e4/remotes/handlers.go#L104). It also uses `content.OpenWriter`, which [internally will check to see if there's already a writer open for that ref](https://github.com/containerd/containerd/blob/334f567e0130b2734a3a5439f2bb644d14aa66e4/content/helpers.go#L98-L99) (and thus that descriptor) and, if so, wait until that writer is closed before opening it again.
3. `fetch` will also always check to see if [opening the writer or committing an already completed writer returns an `AlreadyExists` err](https://github.com/containerd/containerd/blob/334f567e0130b2734a3a5439f2bb644d14aa66e4/remotes/handlers.go#L106-L121), in which case it knows the content was already successfully committed by something else and returns a nil err. Same goes for [the `content.Copy` func it calls internally](https://github.com/containerd/containerd/blob/334f567e0130b2734a3a5439f2bb644d14aa66e4/content/helpers.go#L152).

So, given the above, if we check `isLazy` and it returns true but by the time we actually try to de-lazy it's already been de-lazied by something else, then we should just get a nil error returned by `contentutil.Copy` and can move on successfully. If something else is in the middle of de-lazying, we'll block until that thing is done de-lazying and, if it succeeded, continue on or if it failed, try de-lazying it again.

The only slightly weird corner case I can think of this is if you loaded the same immutable ref twice but with different desc handlers and then those two were both getting de-lazied at the exact same time, it's possible for you to technically pull it from one remote but then the other cache ref wins the race to [setting the `Description` in the metadata](https://github.com/sipsma/buildkit/blob/94e8ab5a3b104d6ba7884375e2d8f391ceb3b1e5/cache/refs.go#L472) and it will be marked there as having been pulled from a different remote image ref. It seems like the description is mostly just informational and if two instances of the same cache ref are being pulled at the same time it is not really significant which one's image ref ends up in the `Description`, but let me know if that matters or if there are other cases I'm not thinking of.

Reply:
We should avoid the case where to parallel `Unlazy` requests make an extra HTTP request to the registry, even if some data is ignored later. Also the locking in `content.Writer` isn't very well thought out, it doesn't handle cancellation of one request gracefully and I believe doesn't even release the lock until a time-based trigger completes. Usually `flightcontrol.Do` that is used extensively in this repo is a much better solution.

Reply:
Ah sure, it should be easy to wrap in a `flightcontrol.Do` too, will do.